### PR TITLE
DOCFIX: Advanced AshPy sections

### LINKED ISSUE #50: [DOC] - Customizing Ashpy Examples
- [ ] Add documentation and examples for the implementation of custom `Metric`.
	- [ ] How to convert a `keras.metrics.Metric` into an Ashpy one?
- [ ] Add documentation and examples for the implementation of custom `Callbacks`.

### PR DESCRIPTION
**Pull request by mr-ubik**
Adds an Advanced AshPy section to the docs with detailing on more advanced uses of the library.

Close #50

### CODE REVIEW

File: docs/source/advanced_ashpy.rst [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+Custom Metrics
+==============
+
+AshPy Trainers can accept metrics that they will use for both logging and automatic model
+selection.
+
+Implementing a custom Metric in AshPy can be done via two approach:
+
+1. Your metrics is already available as a :class:`tf.keras.metrics.Metric` and you want to use it as is.
```
Comment (EmanueleGhelfi) [inline docs/source/advanced_ashpy.rst]: Your Metric I guess

File: docs/source/advanced_ashpy.rst [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+2. You need to write the implementation of the Metric from scratch or you need to alter the default behavior we provide for AshPy Metrics.
+
+Wrapping Keras Metrics
+----------------------
+
+In case number (1) what you want to do is to search for one of the Metrics provided by AshPy
+and use it as a wrapper around the one you wish to use.
+
+.. note::
+    Passing an :mod:`operator` funciton to the AshPy Metric will enable model selection using the
```
Comment (EmanueleGhelfi) [inline docs/source/advanced_ashpy.rst:27]: function

File: docs/source/advanced_ashpy.rst [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            log("generator", out, context.global_step)
+
+Let's start with the :code:`__init__()` function, as for the Custom |ashpy.Metric| when
+inheriting from either |ashpy.Callback| or |ashpy.CounterCallback| respect the common part of the signature:
+
+* :code:`event`: In AshPy we use an Enum - :class:`ashpy.callbacks.Event <ashpy.callbacks.events.Event>` - to
+  choose the event type you want the |ashpy.Callback| to be triggered on.
+* :code:`name`: Unique :obj:`str` identifier for the |ashpy.Callback|
+* :code:`event_freq`: Simple :obj:`int` specifying the frequency.
+* :code:`fn`: A :func:`callable()` this is the funciton that gets triggered. Inside AshPy we
```
Comment (EmanueleGhelfi) [inline docs/source/advanced_ashpy.rst]: function

File: docs/source/advanced_ashpy.rst [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+                out = context.generator_model(context.generator_inputs, training=False)
+            elif context.log_eval_mode == LogEvalMode.TRAIN:
+                out = context.fake_samples
+            else:
+                raise ValueError("Invalid LogEvalMode")
+
+            # tensorboard 2.0 does not support float images in [-1, 1]
+            # only in [0,1]
+            if out.dtype == tf.float32:
+                # The hypothesis is that image are in [-1,1] how to check?
```
Comment (EmanueleGhelfi) [inline docs/source/advanced_ashpy.rst]: Maybe we can remove this

Comment (mr-ubik) [reply docs/source/advanced_ashpy.rst]: I can remove it from the docs but should I remove it also from the code?

Comment (EmanueleGhelfi) [reply docs/source/advanced_ashpy.rst]: No, I guess only from the doc. We should add a flag also for this callback to handle this.
